#637 Add Component Dialog

deanljohnson opened the pull request:
### Summary

Added support for a component search dialog as discussed in #637. The dialog is accessible from the SceneView and from the Object Inspector.

### Details
* Used the previously named ObjectRefSelectionDialog to provide the selection
* Renamed ObjectRefSelectionDialog to SelectionDialog to reflect its more varied use cases
* See image for location and look of buttons added to the Object Inspector

![image](https://user-images.githubusercontent.com/12946946/39493968-8efdf4c6-4d49-11e8-8a28-b21aafc4f9c5.png)

![image](https://user-images.githubusercontent.com/12946946/39494007-ac258078-4d49-11e8-93cf-9a7f030060f5.png)

ilexp commented:
Already looks pretty good 👍 Putting this out for review.

ilexp commented:
Code looks good at a glance, will give it a quick behavior / usability check later and get back to you.

ilexp commented on Source/Editor/DualityEditor/DualityEditor.csproj:
```diff
@@ -121,11 +121,11 @@
     <Compile Include="Extensibility\Previews\IPreviewQuery.cs" />
     <Compile Include="Extensibility\Previews\PreviewSizeMode.cs" />
-    <Compile Include="Forms\ObjectRefSelectionDialog.cs">
+    <Compile Include="Forms\SelectionDialog.cs">
```
Agree 100% to renaming the dialog, but `SelectionDialog` sounds a bit too generic. Not sure what a better name would be, but maybe something like "ObjectSelectionDialog" (without the ref) or "SearchableListSelectionDialog"? Open for suggestions.

ilexp commented on Source/Plugins/EditorBase/PropertyEditors/AddComponentPropertyEditor.cs:
```diff
@@ -0,0 +1,113 @@
+	/// A <see cref="PropertyEditor"/> that provides a button
+	/// for adding a component to a <see cref="GameObject"/>
+	/// </summary>
+	public class AddComponentPropertyEditor : PropertyEditor
```
Works as expected, but for improved usability, the editor should also react to the return key in order to "press" the button - that way the Object Inspector remains usable completely via keyboard. There should be a virtual method you can override to react to keyboard input while the editor is selected.

ilexp commented on Source/Plugins/EditorBase/PropertyEditors/AddComponentPropertyEditor.cs:
```diff
@@ -0,0 +1,113 @@
+
+			this.rectButton = new Rectangle(
+				this.ClientRectangle.X + 2, this.ClientRectangle.Y + 2,
+				this.ClientRectangle.Width - 4, this.ClientRectangle.Height - 4);
```
This is more of a style thing, but can you check how it looks when increasing padding on top and bottom of the editor? Maybe something between 5 and 20 pixels both top and bottom. Ignore if it doesn't turn out to look better.

ilexp commented on Source/Plugins/EditorBase/PropertyEditors/GameObjectOverviewPropertyEditor.cs:
```diff
@@ -120,6 +122,26 @@ protected void UpdateComponentEditors(GameObject[] values)
+			else
+			{
+				this.RemovePropertyEditor(this.addComponentEditor);
+				this.AddPropertyEditor(this.addComponentEditor, this.ChildEditors.Count);
```
Why are you doing the add-remove here? Is it to ensure the editor remains at the bottom as new editors are added? If that's the case, please add a comment above those two lines, so it's clear to the reader.

ilexp commented on Source/Plugins/EditorBase/PropertyEditors/AddComponentPropertyEditor.cs:
```diff
@@ -0,0 +1,113 @@
+		{
+			base.OnMouseMove(e);
+
+			bool buttonContainsMouse = this.rectButton.Contains(e.Location);
```
We don't have this as a use case right now, but to future-proof this, we should already account for `this.ReadOnly`. Readonly should:

- Not react to mouse hover states.
- Not react to keyboard input.
- Display the button in deactivated state.

ilexp commented on Source/Plugins/EditorBase/PropertyEditors/GameObjectPropertyEditor.cs:
```diff
@@ -366,6 +391,34 @@ private void OnPrefabLinkBreakPressed()
+			// state caused by the SelectionDialog being modal
+			this.curButton = -1;
+			this.curButtonHovered = false;
+			this.curButtonPressed = false;
```
👍 for the explanatory comment and dealing with that case.

ilexp commented on Source/Plugins/EditorModules/SceneView/Modules/SceneView.cs:
```diff
@@ -678,9 +678,16 @@ protected void InitContextMenu()
 						},
 						new MenuModelItem
+						{
+							Name            = typeof(Component).Name,
```
Can we add a "..." here to indicate that selecting this menu item will prompt a dialog for user input, rather than immediately performing an action? Would clarify what this item does a bit more by sticking to the "Windows convention".

deanljohnson commented:
I've addressed your points except the name of the dialog:

ObjectRefSelectionDialog: doesn't work because it is not generic enough and doesn't always select references
ObjectSelectionDialog: my issue with this one is that it does not necessarily select objects (unless we are talking objects as in .NET objects, but that is just too generic).
SearchableListSelectionDialog: I feel like the fact that it is searchable is a detail not need in the name
SelectionDialog: too generic as to the type of selection taking place

How about ListSelectionDialog?

ilexp commented:
Good points for all the names. Let's go with ListSelectionDialog for now.

ilexp commented:
Will check out the new version 👍

ilexp commented:
Merged. Thanks! 👍
